src/v/security/gssapi_principal_mapper.cc

```diff
@@ -0,0 +1,505 @@
+        auto base = replace_parameters(_format, params);
+        if (!base) {
+            return std::nullopt;
+        }
+        const re2::StringPiece base_piece(base->data(), base->size());
+        if (_match.empty() || re2::RE2::FullMatch(base_piece, match_regex)) {
+            if (!_from_pattern) {
+                result = *base;
+            } else {
+                result = replace_substitution(
+                  *base, *_from_pattern, _to_pattern, _repeat);
+            }
+        }
+    }
+
+    const re2::StringPiece result_piece(result.data(), result.size());
+    if (
+      !result.empty()
+      && re2::RE2::PartialMatch(result_piece, non_simple_regex)) {
+        vlog(
+          seclog.error,
+          "Non-simple name {} after auth_to_local rule {}",
+          result,
+          *this);
+        return std::nullopt;
+    }
+
+    if (!result.empty()) {
+        switch (_case_change) {
+        case case_change_operation::noop:
+            break;
+
+        case case_change_operation::make_lower:
+            boost::algorithm::to_lower(result, std::locale::classic());
+            break;
+
+        case case_change_operation::make_upper:
+            boost::algorithm::to_upper(result, std::locale::classic());
+            break;
+        }
+    }
+
+    if (result.empty()) {
+        return std::nullopt;
+    } else {
+        return result;
+    }
+}
+
+std::optional<ss::sstring> gssapi_rule::replace_parameters(
+  std::string_view format, std::vector<std::string_view> params) {
+    static thread_local const re2::RE2 parameter_parser(
+      parameter_pattern, re2::RE2::Quiet);
+    vassert(
+      parameter_parser.ok(),
+      "Invalid parameter pattern: {}",
+      parameter_parser.error());
+
+    ss::sstring result;
+    re2::StringPiece format_piece(format.data(), format.size());
+
+    re2::StringPiece text_replace, index_replace;
+    while (re2::RE2::Consume(
+      &format_piece,
+      parameter_parser,
+      &text_replace,
+      nullptr,
+      &index_replace)) {
+        if (!text_replace.empty()) {
+            result.append(text_replace.data(), text_replace.size());
+        }
+        if (!index_replace.empty()) {
+            std::size_t index = std::numeric_limits<std::size_t>::max();
+            auto conv_result = std::from_chars(
+              index_replace.begin(), index_replace.end(), index);
+            if (conv_result.ec != std::errc()) {
+                vlog(
+                  seclog.warn,
+                  "Bad format in username mapping in {}",
+                  std::string_view{index_replace.data(), index_replace.size()});
+                return std::nullopt;
+            } else if (index >= params.size()) {
+                vlog(
+                  seclog.warn,
+                  "Invalid index provided ({}).  Outside range of "
+                  "parameters (length {})",
+                  index,
+                  params.size());
```

Comment:
Warning -> User provided a bad rule to us but is 'technically' valid.

Reply:
i don't have a good intuition about log levels in general, but im wondering why warning--it doesn't seem like redpanda itself is behaving in a concerning way, right?